### PR TITLE
[Feature] compaction quickly for small data import #9791

### LINKED ISSUE #9791: [Feature] compaction quickly for samll data import
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-doris/issues?q=is%3Aissue) and found no similar issues.

### Description

_No response_

### Use case

_No response_

### Related issues

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by chenlinzhong**
# Proposed changes

Issue Number: close #9791

## Problem Summary:
A table frequently imports a small amount of data. Doris generates a version for each imported data. If these versions are not merged in time, two problems will occur

- -235: too many versions, exceeding the maximum number of tablet versions, import failed
- query performance: doris adopts the merge on read mechanism, all versions of data will be merged when reading data. Too many versions may affect query performance

For this reason, we hope to merge the imported versions faster. The simplest way is to increase the number of CC pool thread numbers . However, this may lead to too high io of the entire node, because the CC task may select rowsets with a large amount of data

So We hope to merge some rowsets as soon as possible without adding disk IO

Therefore, we choose rowset with small data rowset to merge first. The definition of small data rowset is that the imported rows are smaller than config:: quick compaction_ max_ rows

To achieve this goal
we add a new  thread pool for  small rowset  compaction. This thread pool only selects rowsets with small data rowsets for merging. The time to trigger merging is as follows:

- Import completed: when be receives the push version request,  try to select a small version of rowset. when the number of rowsets of selected exceeds config:: quick_compaction_ min_ rowsets do a compaction
- -235: try to trigger when -235 appears

### config  parameter
| key | desc | default |
| :-----| :---- | :----: |
| config::enable_quick_compaction |  whether to  turn on this feature | false |
| config::quick_compaction_max_rows | less then rows rowset will be selected  | 1000 |
| config::quick_compaction_batch_size | triggered every publish batch size    | 10 |
| config::quick_compaction_min_rowsets | do quick compaction min selected rowsets | 10 |

#### 1.we create a table, try to see the quick compaction cost of different rows with centern rowsets
```
CREATE TABLE table1 (
event_day date NULL COMMENT "",
siteid int(11) NULL DEFAULT "10" COMMENT "",
citycode smallint(6) NULL COMMENT "",
username varchar(32) NULL DEFAULT "" COMMENT "",
pv bigint(20) SUM NULL DEFAULT "0" COMMENT ""
) ENGINE=OLAP
AGGREGATE KEY(event_day, siteid, citycode, username)
COMMENT "OLAP"
DISTRIBUTED BY HASH(username) BUCKETS 10
PROPERTIES (
"replication_allocation" = "tag.location.default: 1",
"in_memory" = "false",
"storage_format" = "V2"
)
```
here is the result
image

#### 2. test import small data performance with high frequency 

start 30 threads ,every thread import  100 batch, every batch included 500-1000 rows, the result:

image

image

image

image

image

we can concluded that 
- cc compaction can not compact rowset in time and will stop import process when reach the max versions limits
- quick compaction can effectively handle  small data import  scene and compact rowset in time

#### 3.test one be import max qps with small data  and compare with quick compaction
[without quick compaction]
start 2 threads ,every thread import  100 batch, every batch included 500-1000 rows
image
image
we can concluded that 

- max_qps is about 22/s,
- average qps is about 10-15/s 
- import will usually break by -235 

[with quick compaction]
start 30 threads ,every thread import  1000 batch, every batch included 500-1000 rows
image
image
image

we can concluded that 
- max_qps is about 32/s
- average qps is 30/s
- import will never break by -235  and very stable
- disk IO remains low level 

### conclusion

- quick compaction can effectively solve -235 problems 
- quick compaction can imporve  import qps more than three times 
- quick compaction will not  cost a lot of extra disk io
- quick compaction is suitable for scenarios with small amount of data import

## Checklist(Required)

1. Does it affect the original behavior: (Yes/No/I Don't know)
2. Has unit tests been added: (Yes/No/No Need)
4. Has document been added or modified: (Yes/No/No Need)
5. Does it need to update dependencies: (Yes/No)
6. Are there any changes that cannot be rolled back: (Yes/No)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by morningman** [comment]
Could you provide some test result for this feature?

File: be/src/olap/tablet.cpp [hunk 1]
```diff
@@ -839,10 +839,53 @@ void Tablet::calculate_cumulative_point() {
     if (ret_cumulative_point == K_INVALID_CUMULATIVE_POINT) {
         return;
     }
-
     set_cumulative_layer_point(ret_cumulative_point);
 }
 
+//1.小版本rowset
+//2.不改变cp，cp值的更新有cc任务来执行
+//3.找到最大的，连续小版本rowset
```
Comment (Gabriel39) [inline be/src/olap/tablet.cpp:847]: Use English comments pls

**Comment by chenlinzhong** [comment]
> Could you provide some test result for this feature?

ok

File: be/src/olap/tablet.cpp (hunk 1 above)
Comment (chenlinzhong) [reply be/src/olap/tablet.cpp:847]: > Use English comments pls

ok

File: be/src/vec/common/cow.h [hunk 2]
```diff
@@ -350,8 +350,8 @@ class COW {
         const T& operator*() const { return *value; }
         T& operator*() { return value->assume_mutable_ref(); }
 
-        operator const immutable_ptr<T>&() const { return value; }
-        operator immutable_ptr<T>&() { return value; }
```
Comment (BiteTheDDDDt) [inline be/src/vec/common/cow.h:354]: Maybe you should update your formatter. https://github.com/apache/incubator-doris/pull/9602

File: be/src/olap/compaction.cpp [hunk 3]
```diff
@@ -55,6 +55,25 @@ Status Compaction::execute_compact() {
...
+    std::unique_lock<std::mutex> lock(_tablet->get_cumulative_compaction_lock(), std::try_to_lock);
+    if (!lock.owns_lock()) {
+        LOG(INFO) << "The tablet is under cumulative compaction. tablet=" << _tablet->full_name();
+        return Status::OLAPInternalError(OLAP_ERR_CE_TRY_CE_LOCK_ERROR);
+    }
+    _input_rowsets.clear();
+    int vertion_count = _tablet->version_count();
+    int64_t now = UnixMillis();
+    _tablet->pick_samll_verson_rowsets(&_input_rowsets);
+    if (_input_rowsets.size() >= 5) {
```
Comment (BiteTheDDDDt) [inline be/src/olap/compaction.cpp:68]: Define a constant in place of the magic number

File: be/src/vec/common/cow.h (hunk 2 above)
Comment (chenlinzhong) [reply be/src/vec/common/cow.h:354]: ok

File: be/src/olap/compaction.cpp (hunk 3 above)
Comment (chenlinzhong) [reply be/src/olap/compaction.cpp:68]: ok

File: be/src/olap/tablet.cpp [hunk 4]
```diff
@@ -839,10 +839,53 @@ void Tablet::calculate_cumulative_point() {
...
+    int max_series_num = 1000;
+    int max_rows = config::small_version_max_rows;
+    if (max_rows <= 0) return Status::OK();
+    std::vector<std::vector<RowsetSharedPtr>> samll_version_rowsets(max_series_num);
+    int idx = 0;
+    bool is_bad = false;
+    if (tablet_state() == TABLET_RUNNING) {
+        for (auto& rs : _rs_version_map) {
+            bool is_delete = version_for_delete_predicate(rs.first);
+            if (rs.first.first > cumulative_layer_point()) {
```
Comment (BiteTheDDDDt) [inline be/src/olap/tablet.cpp:858]: Maybe you can use more `guard clauses` on here.

File: be/src/olap/compaction.cpp [hunk 5]
```diff
@@ -55,6 +55,25 @@ Status Compaction::execute_compact() {
     return st;
 }
 
+Status Compaction::samll_rowsets_compact() {
```
Comment (caiconghui) [inline be/src/olap/compaction.cpp:58]:
```suggestion
Status Compaction::small_rowsets_compact() {
```

Comment (yinzhijian) [inline be/src/olap/compaction.cpp:58]: small

File: be/src/olap/tablet.cpp [hunk 6]
```diff
@@ -839,10 +839,53 @@ void Tablet::calculate_cumulative_point() {
...
         return;
     }
-
     set_cumulative_layer_point(ret_cumulative_point);
 }
 
+//1.小版本rowset
+//2.不改变cp，cp值的更新有cc任务来执行
+//3.找到最大的，连续小版本rowset
+Status Tablet::pick_samll_verson_rowsets(std::vector<RowsetSharedPtr>* input_rowsets) {
```
Comment (yinzhijian) [inline be/src/olap/tablet.cpp:848]: small_version

File: be/src/olap/olap_server.cpp [hunk 7]
```diff
@@ -72,6 +72,11 @@ Status StorageEngine::start_bg_threads() {
             .set_max_threads(max_thread_num)
             .build(&_compaction_thread_pool);
 
+    ThreadPoolBuilder("CompactionTaskThreadPool")
+            .set_min_threads(max_thread_num)
```
Comment (yinzhijian) [inline be/src/olap/olap_server.cpp:76]: add a new config max_small_compaction_threads for finer control
?

File: be/src/olap/compaction.h [hunk 8]
```diff
@@ -48,6 +48,7 @@ class Compaction {
 
     // This is only for http CompactionAction
     Status compact();
+    Status samll_rowsets_compact();
```
Comment (yinzhijian) [inline be/src/olap/compaction.h:51]: If BaseCompaction.samll_rowsets_compact() is called, what is its behavior?

File: be/src/olap/compaction.cpp [hunk 9]
```diff
@@ -55,6 +55,25 @@ Status Compaction::execute_compact() {
...
+    std::unique_lock<std::mutex> lock(_tablet->get_cumulative_compaction_lock(), std::try_to_lock);
+    if (!lock.owns_lock()) {
+        LOG(INFO) << "The tablet is under cumulative compaction. tablet=" << _tablet->full_name();
+        return Status::OLAPInternalError(OLAP_ERR_CE_TRY_CE_LOCK_ERROR);
+    }
+    _input_rowsets.clear();
+    int vertion_count = _tablet->version_count();
```
Comment (Gabriel39) [inline be/src/olap/compaction.cpp:65]:
```suggestion
    int version_count = _tablet->version_count();
```

File: be/src/olap/compaction.cpp [hunk 10]
```diff
@@ -55,6 +55,25 @@ Status Compaction::execute_compact() {
...
+    int64_t now = UnixMillis();
+    _tablet->pick_samll_verson_rowsets(&_input_rowsets);
+    if (_input_rowsets.size() >= 5) {
+        do_compaction(0);
+        LOG(INFO) << "samll_rowsets_compact,before_versions:" << vertion_count
```
Comment (Gabriel39) [inline be/src/olap/compaction.cpp:70]:
```suggestion
        LOG(INFO) << "samll_rowsets_compact,before_versions:" << version_count
```

File: be/src/olap/tablet.cpp [hunk 11]
```diff
@@ -839,10 +839,53 @@ void Tablet::calculate_cumulative_point() {
...
+    int max_series_num = 1000;
+    int max_rows = config::small_version_max_rows;
+    if (max_rows <= 0) return Status::OK();
+    std::vector<std::vector<RowsetSharedPtr>> samll_version_rowsets(max_series_num);
+    int idx = 0;
+    bool is_bad = false;
```
Comment (Gabriel39) [inline be/src/olap/tablet.cpp:854]: Unused variable?

File: be/src/olap/tablet.cpp [hunk 12]
```diff
@@ -839,10 +839,53 @@ void Tablet::calculate_cumulative_point() {
...
+            }
+        }
+        int max_idx = samll_version_rowsets.size() > 0 ? 0 : -1;
+        std::vector<RowsetSharedPtr> result;
+        if (is_bad) Status::OK();
+        for (int i = 0; i < samll_version_rowsets.size(); i++) {
+            if (samll_version_rowsets[i].size() > samll_version_rowsets[max_idx].size()) {
+                max_idx = i;
+            }
+            result = samll_version_rowsets[max_idx];
```
Comment (Gabriel39) [inline be/src/olap/tablet.cpp:879]: seems you should move this line into the above `if` code block

File: be/src/olap/tablet.cpp [hunk 13]
```diff
@@ -839,10 +839,53 @@ void Tablet::calculate_cumulative_point() {
...
+                    } else {
+                        idx++;
+                        if (idx > max_series_num) {
+                            break;
+                        }
+                    }
+                }
+            }
+        }
+        int max_idx = samll_version_rowsets.size() > 0 ? 0 : -1;
```
Comment (Gabriel39) [inline be/src/olap/tablet.cpp:872]: Just set `max_idx` to 0

File: be/src/olap/delta_writer.cpp [hunk 14]
```diff
@@ -98,6 +100,8 @@ Status DeltaWriter::init() {
             MemTracker::create_tracker(-1, "DeltaWriter:" + std::to_string(_tablet->tablet_id()));
     // check tablet version number
     if (_tablet->version_count() > config::max_tablet_version_num) {
+        //trigger samll compaction
+        StorageEngine::instance()->submit_samll_compaction_task(_tablet);
```
Comment (Gabriel39) [inline be/src/olap/delta_writer.cpp:104]: should we check whether to pass this check after small compaction?

File: be/src/agent/task_worker_pool.cpp [hunk 15]
```diff
@@ -717,7 +719,18 @@ void TaskWorkerPool::_publish_version_worker_thread_callback() {
                          << ", error_code=" << res;
             finish_task_request.__set_error_tablet_ids(error_tablet_ids);
         } else {
-            LOG(INFO) << "publish_version success. signature:" << agent_task_req.signature;
+            for (int i = 0; i < succ_tablet_ids.size(); i++) {
+                TabletSharedPtr tablet =
+                        StorageEngine::instance()->tablet_manager()->get_tablet(succ_tablet_ids[i]);
+                if (tablet != nullptr) {
```
Comment (Gabriel39) [inline be/src/agent/task_worker_pool.cpp:725]: should we always do small compaction after PublishVersionTask?

File: be/src/common/config.h [hunk 16]
```diff
@@ -743,6 +743,9 @@ CONF_Int32(object_pool_buffer_size, "100");
 // ParquetReaderWrap prefetch buffer size
 CONF_Int32(parquet_reader_max_buffer_size, "50");
 
+// For continuous versions that rows less than small_version_max_rows will  trigger compaction quickly
+CONF_Int32(small_version_max_rows, "1000");
```
Comment (yangzhg) [inline be/src/common/config.h:747]: You'd better to add a config to turn on/off this feature

File: be/src/agent/task_worker_pool.cpp [hunk 17]
```diff
@@ -717,7 +719,18 @@ void TaskWorkerPool::_publish_version_worker_thread_callback() {
                          << ", error_code=" << res;
             finish_task_request.__set_error_tablet_ids(error_tablet_ids);
         } else {
-            LOG(INFO) << "publish_version success. signature:" << agent_task_req.signature;
+            for (int i = 0; i < succ_tablet_ids.size(); i++) {
+                TabletSharedPtr tablet =
+                        StorageEngine::instance()->tablet_manager()->get_tablet(succ_tablet_ids[i]);
```
Comment (yangzhg) [inline be/src/agent/task_worker_pool.cpp:724]: It's better to submit small version compaction in batch, for example  ervery 20 small verison

File: be/src/common/config.h (hunk 16 above)
Comment (chenlinzhong) [reply be/src/common/config.h:747]: > You'd better to add a config to turn on/off this feature

set 0 means turn off this feature

File: be/src/olap/compaction.cpp (hunk 5 above)
Comment (chenlinzhong) [reply be/src/olap/compaction.cpp]: fix

Comment (chenlinzhong) [reply be/src/olap/compaction.cpp]: > small

fixed

File: be/src/olap/tablet.cpp (hunk 6 above)
Comment (chenlinzhong) [reply be/src/olap/tablet.cpp]: > small_version

fixed

File: be/src/olap/compaction.cpp (hunk 9 above)
Comment (chenlinzhong) [reply be/src/olap/compaction.cpp]: > 

fixed

File: be/src/olap/compaction.cpp (hunk 10 above)
Comment (chenlinzhong) [reply be/src/olap/compaction.cpp]: > 

fixed

File: be/src/olap/delta_writer.cpp (hunk 14 above)
Comment (chenlinzhong) [reply be/src/olap/delta_writer.cpp]: > should we check whether to pass this check after small compaction?

Small version merging is asynchronous, which is unlikely to be completed in a short time, so it is not necessary to check here

File: be/src/agent/task_worker_pool.cpp (hunk 17 above)
Comment (chenlinzhong) [reply be/src/agent/task_worker_pool.cpp:725]: > It's better to submit small version compaction in batch, for example ervery 20 small verison

ok ,done

File: be/src/agent/task_worker_pool.cpp (hunk 15 above)
Comment (chenlinzhong) [reply be/src/agent/task_worker_pool.cpp:729]: > should we always do small compaction after PublishVersionTask?

 not necessary ，changed to  batch submit

File: be/src/common/config.h [hunk 18]
```diff
@@ -300,6 +300,9 @@ CONF_mInt32(convert_rowset_thread_num, "0");
 // initial sleep interval in seconds of scan alpha rowset
 CONF_mInt32(scan_alpha_rowset_min_interval_sec, "3");
 
+// This config can be set to limit thread number in  smallcompaction thread pool.
+CONF_mInt32(small_compaction_max_threads, "10");
```
Comment (yangzhg) [inline be/src/common/config.h:304]: use quick compaction is better

File: be/src/olap/compaction.cpp [hunk 19]
```diff
@@ -55,6 +55,68 @@ Status Compaction::execute_compact() {
...
+        _tablet->set_clone_occurred(false);
+        return Status::OLAPInternalError(OLAP_ERR_CUMULATIVE_CLONE_OCCURRED);
+    }
+
+    _input_rowsets.clear();
+    int version_count = _tablet->version_count();
+    int64_t now = UnixMillis();
+    int64_t permits = 0;
+    _tablet->pick_small_verson_rowsets(&_input_rowsets, &permits);
+    std::string input_ver = "";
```
Comment (yangzhg) [inline be/src/olap/compaction.cpp:78]: use sstream, and Version has implement operator <<

File: be/src/olap/compaction.cpp [hunk 20]
```diff
@@ -55,6 +55,68 @@ Status Compaction::execute_compact() {
...
+    // Clone task may happen after compaction task is submitted to thread pool, and rowsets picked
+    // for compaction may change. In this case, current compaction task should not be executed.
+    if (_tablet->get_clone_occurred()) {
+        _tablet->set_clone_occurred(false);
+        return Status::OLAPInternalError(OLAP_ERR_CUMULATIVE_CLONE_OCCURRED);
+    }
+
+    _input_rowsets.clear();
+    int version_count = _tablet->version_count();
+    int64_t now = UnixMillis();
```
Comment (yangzhg) [inline be/src/olap/compaction.cpp:75]: you can use MonotonicStopWatch

File: be/src/olap/tablet.cpp [hunk 21]
```diff
@@ -870,10 +870,56 @@ void Tablet::calculate_cumulative_point() {
...
 
+//find rowsets that rows less then "config::small_compaction_max_rows"
+Status Tablet::pick_small_verson_rowsets(std::vector<RowsetSharedPtr>* input_rowsets,
+                                         int64_t* permits) {
+    if (!init_succeeded()) {
+        return Status::OLAPInternalError(OLAP_ERR_CUMULATIVE_INVALID_PARAMETERS);
+    }
+    int max_series_num = 1000;
+    int max_rows = config::small_compaction_max_rows;
+    if (max_rows <= 0) return Status::OK();
```
Comment (yangzhg) [inline be/src/olap/tablet.cpp:884]:
```suggestion
    if (max_rows <= 0) { return Status::OK(); }
```

**Comment by yangzhg** [comment]
Can you provide some test data about how often this feature can be imported with and without it

**Comment by caiconghui** [comment]
will there be any side effect for normal compaction performance, if there are many stream load and with not small data?

**Comment by chenlinzhong** [comment]
> will there be any side effect for normal compaction performance, if there are many stream load and with not small data?

maybe， because  small compaction use the same lock with cc compaction, so small compaction may stop  cc compaction for one tablet at this round , but small compaction finished very quickly usually less then 1s , cc compaction maybe excute next round

**Comment by caiconghui** [comment]
> > will there be any side effect for normal compaction performance, if there are many stream load and with not small data?
> 
> maybe， because small compaction use the same lock with cc compaction, so small compaction may stop cc compaction for one tablet at this round , but small compaction finished very quickly usually less then 1s , cc compaction maybe excute next round

so，it should be optional choice for user to use this optimization, may be better?

**Review by yangzhg — APPROVED** [review]
LGTM

File: be/src/olap/tablet.cpp [hunk 22]
```diff
@@ -870,10 +870,59 @@ void Tablet::calculate_cumulative_point() {
...
+        if (quick_compaction_rowsets.size() == 0) return Status::OK();
+        std::vector<RowsetSharedPtr> result = quick_compaction_rowsets[0];
+        for (int i = 0; i < quick_compaction_rowsets.size(); i++) {
+            if (quick_compaction_rowsets[i].size() > result.size()) {
+                result = quick_compaction_rowsets[i];
+            }
+        }
+        for (int i = 0; i < result.size(); i++) {
+            *permits += result[i]->num_segments();
+            input_rowsets->push_back(result[i]);
```
[the remark this answers is not in the document]
Comment (chenlinzhong) [reply be/src/olap/tablet.cpp:920]: check consecutive  later in  Compaction::quick_rowsets_compact ,if not will give up

**Comment by chenlinzhong** [comment]
> will there be any side effect for normal compaction performance, if there are many stream load and with not small data?

I think it will not , the quick compaciton will do nothing if not with samll data

**Pull request merged**